File: sdks/typescript/src/governed-change.ts
import { createHash, createHmac } from "node:crypto";
import type {
  AuditEventInput,
  EvidenceEdgeInput,
  EvidenceEntity,
  EvidenceScope,
  JsonObject,
  JsonValue,
  Principal,
} from "./index";
import { canonicalJson } from "./index";

export type EvidenceRefKind = "principal" | "entity" | "activity" | "change" | "revision" | "assertion" | "record" | "commit";

export interface EvidenceRef {
  authority: string;
  kind: EvidenceRefKind;
  type: string;
  id: string;
}

export interface CapturePolicyRef {
  id: string;
  version: string;
}

export type CaptureMode =
  | "omit"
  | "content_digest"
  | "keyed_digest"
  | "randomized_digest"
  | "reference"
  | "redact"
  | "encrypt"
  | "full";

export interface EntityFieldPolicy {
  capture: CaptureMode;
}

export interface GovernedEntityDefinition<Row extends Record<string, unknown>> {
  authority: string;
  type: string;
  schemaRef: string;
  fieldSetRef: string;
  identity(row: Row): string;
  fields: Partial<Record<keyof Row & string, EntityFieldPolicy>>;
  lineagePolicy?: "linear" | "dag";
}

export interface GovernedEntity<Row extends Record<string, unknown>> extends GovernedEntityDefinition<Row> {
  ref(rowOrId: Row | string): EvidenceRef;
}

export interface VeritioContextMetadata {
  authSessionId?: string;
  authContextId?: string;
  activityEpisodeId?: string;
  traceId?: string;
  correlationId?: string;
  causationEventId?: string;
  changeId?: string;
  capturePolicyId?: string;
  collectionSource?: string;
}

export interface StateCommitment {
  algorithm: "sha256";
  canonicalization: "veritio-json-v1";
  schemaRef: string;
  fieldSetRef: string;
  fields: JsonObject;
  digest: string;
}

export interface RevisionDraft {
  ref: EvidenceRef;
  entity: EvidenceRef;
  parents: EvidenceRef[];
  stateCommitment: StateCommitment;
  changedPaths: string[];
  generatedBy: EvidenceRef;
  capturePolicyRef?: CapturePolicyRef;
}

export interface GovernedChangeDraftInput<Row extends Record<string, unknown>> {
  scope: EvidenceScope & { tenantId: string };
  entity: GovernedEntity<Row>;
  before?: Row;
  after: Row;
  changedPaths: string[];
  change: {
    id: string;
    type: string;
    initiatedBy: EvidenceRef;
    authorizationAssertionRef?: EvidenceRef;
    delegationAssertionRef?: EvidenceRef;
  };
  activity: {
    id: string;
    type: string;
    performedBy: EvidenceRef;
  };
  producer: EvidenceRef;
  occurredAt: string | Date;
  idempotencyKeyHash: string;
  context?: VeritioContextMetadata;
  metadata?: Record<string, unknown>;
  capturePolicyRef?: CapturePolicyRef;
  expectedParentRevisionRef?: EvidenceRef;
  mutationBinding?: "same_transaction" | "not_transaction_bound" | "best_effort";
  digestKeys?: {
    keyedDigest?: {
      keyVersion: string;
      secret: string;
    };
  };
}

export interface GovernedChangeDraft {
  changeRef: EvidenceRef;
  activityRef: EvidenceRef;
  entityRef: EvidenceRef;
  revision: RevisionDraft;
  events: AuditEventInput[];
  edges: EvidenceEdgeInput[];
  outboxEntry: {
    schemaVersion: "2026-06-23";
    mutationBinding: "same_transaction" | "not_transaction_bound" | "best_effort";
    expectedParentRevisionRef?: EvidenceRef;
    records: AuditEventInput[];
    edges: EvidenceEdgeInput[];
  };
}

const RESERVED_CONTEXT_KEYS = [
  "authSessionId",
  "authContextId",
  "activityEpisodeId",
  "traceId",
  "correlationId",
  "causationEventId",
  "changeId",
  "capturePolicyId",
  "collectionSource",
] as const;

/**
 * Formats an authority-qualified evidence reference for deterministic maps,
 * route keys, and logs without dropping the authority that makes the ID safe to
 * join across systems.
 */
export function refKey(ref: EvidenceRef): string {
  assertRef(ref);
  return `${ref.authority}:${ref.kind}:${ref.type}:${ref.id}`;
}

/**
 * Registers a governed entity type at the host boundary. The SDK stores only the
 * declaration needed to derive references and minimized revision commitments; it
 * does not read application databases or environment state.
 */
export function defineEntity<Row extends Record<string, unknown>>(
  definition: GovernedEntityDefinition<Row>,
): GovernedEntity<Row> {
  assertNonEmpty(definition.authority, "authority");
  assertNonEmpty(definition.type, "type");
  assertNonEmpty(definition.schemaRef, "schemaRef");
  assertNonEmpty(definition.fieldSetRef, "fieldSetRef");

  return {
    ...definition,
    /**
     * Builds the stable authority-qualified entity ref for a row or already
     * resolved host ID.
     */
    ref(rowOrId) {
      const id = typeof rowOrId === "string" ? rowOrId : definition.identity(rowOrId);
      assertNonEmpty(id, "entity.id");
      return {
        authority: definition.authority,
        kind: "entity",
        type: definition.type,
        id,
      };
    },
  };
}

/**
 * Merges caller metadata with SDK-owned context keys. Reserved context keys are
 * applied after caller metadata and callers cannot shadow them, preserving the
 * grouping semantics used by Change, Trace, and Explain projections.
 */
export function mergeVeritioMetadata(
  callerMetadata: Record<string, unknown> = {},
  context: VeritioContextMetadata = {},
): Record<string, unknown> {
  for (const key of RESERVED_CONTEXT_KEYS) {
    if (Object.hasOwn(callerMetadata, key)) {
      throw new TypeError(`metadata.${key} is reserved by Veritio`);
    }
  }

  const merged: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(callerMetadata).sort()) {
    if (value !== undefined && value !== null) {
      merged[key] = value;
    }
  }
  for (const key of RESERVED_CONTEXT_KEYS) {
    const value = context[key];
    if (value !== undefined) {
      merged[key] = value;
    }
  }
  return merged;
}

/**
 * Creates a current-protocol governed-change draft. The draft uses normal audit
 * events plus evidence edges so v1 stores can append it without claiming
 * EvidenceCommit atomicity before that protocol exists.
 */
export function createGovernedChangeDraft<Row extends Record<string, unknown>>(
  input: GovernedChangeDraftInput<Row>,
): GovernedChangeDraft {
  assertNonEmpty(input.scope.tenantId, "scope.tenantId");
  assertRef(input.change.initiatedBy);
  assertRef(input.activity.performedBy);
  assertRef(input.producer);

  const occurredAt = normalizeDate(input.occurredAt);
  const entityRef = input.entity.ref(input.after);
  const changeRef: EvidenceRef = {
    authority: "veritio",
    kind: "change",
    type: input.change.type,
    id: input.change.id,
  };
  const activityRef: EvidenceRef = {
    authority: "veritio",
    kind: "activity",
    type: input.activity.type,
    id: input.activity.id,
  };
  if (input.expectedParentRevisionRef) {
    assertRef(input.expectedParentRevisionRef);
  }
  const previousRevisionRef: EvidenceRef =
    input.expectedParentRevisionRef ?? {
      authority: "veritio",
      kind: "revision",
      type: input.entity.type,
      id: `rev_${input.entity.type}_${entityRef.id}_previous`,
    };
  const stateCommitment = createStateCommitment(input.entity, input.after, input.digestKeys);
  const revisionRef: EvidenceRef = {
    authority: "veritio",
    kind: "revision",
    type: input.entity.type,
    id: `rev_${input.entity.type}_${entityRef.id}_${stateCommitment.digest.slice("sha256:".length, "sha256:".length + 12)}`,
  };
  const revision: RevisionDraft = {
    ref: revisionRef,
    entity: entityRef,
    parents: input.before ? [previousRevisionRef] : [],
    stateCommitment,
    changedPaths: [...input.changedPaths].sort(),
    generatedBy: activityRef,
  };
  if (input.capturePolicyRef) {
    revision.capturePolicyRef = input.capturePolicyRef;
  }

  const metadata = mergeVeritioMetadata(input.metadata, input.context);
  const captureAssurance = {
    captureMethod: "transactional_outbox",
    mutationBinding: input.mutationBinding ?? "not_transaction_bound",
  };
  const common = {
    scope: input.scope,
    occurredAt,
    purpose: "change_provenance",
    dataCategories: ["source_reference"],
    retention: "change_1y",
  };
  const events: AuditEventInput[] = [
    {
      ...common,
      id: `evt_change_declared_${input.change.id}`,
      actor: principalFromRef(input.change.initiatedBy),
      action: "change.declared",
      target: { type: "change", id: input.change.id },
      metadata: compactMetadata({
        ...metadata,
        recordType: "change.declared",
        recordAuthority: changeRef.authority,
        producer: input.producer,
        initiatedBy: input.change.initiatedBy,
        changeType: input.change.type,
        idempotencyKeyHash: input.idempotencyKeyHash,
        capturePolicyRef: input.capturePolicyRef,
        authorizationAssertionRef: input.change.authorizationAssertionRef,
        delegationAssertionRef: input.change.delegationAssertionRef,
        captureAssurance,
      }),
    },
    {
      ...common,
      id: `evt_activity_recorded_${input.activity.id}`,
      actor: principalFromRef(input.activity.performedBy),
      action: "activity.recorded",
      target: { type: "activity", id: input.activity.id },
      metadata: compactMetadata({
        ...metadata,
        recordType: "activity.recorded",
        recordAuthority: activityRef.authority,
        producer: input.producer,
        performedBy: input.activity.performedBy,
        activityType: input.activity.type,
        idempotencyKeyHash: input.idempotencyKeyHash,
        captureAssurance,
      }),
    },
    {
      ...common,
      id: `evt_entity_revision_${revisionRef.id}`,
      actor: principalFromRef(input.producer),
      action: "entity.revision.created",
      target: { type: input.entity.type, id: entityRef.id },
      metadata: compactMetadata({
        ...metadata,
        recordType: "entity.revision",
        recordAuthority: revisionRef.authority,
        producer: input.producer,
        idempotencyKeyHash: input.idempotencyKeyHash,
        veritio: { revision },
        captureAssurance,
      }),
    },
  ];

  const edges: EvidenceEdgeInput[] = [
    draftEdge("has_activity", changeRef, activityRef, occurredAt, input.scope),
    draftEdge("has_output", changeRef, revisionRef, occurredAt, input.scope),
    draftEdge("performed_by", activityRef, input.activity.performedBy, occurredAt, input.scope),
    draftEdge("generated", activityRef, revisionRef, occurredAt, input.scope),
  ];
  if (input.before) {
    edges.push(draftEdge("derived_from", revisionRef, previousRevisionRef, occurredAt, input.scope));
  }

  const outboxEntry: GovernedChangeDraft["outboxEntry"] = {
    schemaVersion: "2026-06-23",
    mutationBinding: input.mutationBinding ?? "not_transaction_bound",
    records: events,
    edges,
  };
  if (input.before) {
    outboxEntry.expectedParentRevisionRef = previousRevisionRef;
  }

  return {
    changeRef,
    activityRef,
    entityRef,
    revision,
    events,
    edges,
    outboxEntry,
  };
}

/**
 * Builds the revision state commitment after applying the entity field policy.
 * Default handling intentionally stores no undeclared fields so the outbox cannot
 * become a raw copy of application rows.
 */
function createStateCommitment<Row extends Record<string, unknown>>(
  entity: GovernedEntity<Row>,
  row: Row,
  digestKeys: GovernedChangeDraftInput<Row>["digestKeys"] = {},
): StateCommitment {
  const fields: JsonObject = {};
  for (const key of Object.keys(entity.fields).sort()) {
    const policy = entity.fields[key];
    if (!policy || policy.capture === "omit") {
      continue;
    }
    const value = row[key];
    if (value === undefined) {
      continue;
    }
    if (policy.capture === "full") {
      fields[key] = toJsonValue(value);
    } else if (policy.capture === "keyed_digest") {
      const keyedDigest = digestKeys.keyedDigest;
      if (!keyedDigest) {
        throw new TypeError("digestKeys.keyedDigest is required for keyed_digest fields");
      }
      assertNonEmpty(keyedDigest.keyVersion, "digestKeys.keyedDigest.keyVersion");
      assertNonEmpty(keyedDigest.secret, "digestKeys.keyedDigest.secret");
      fields[key] = {
        algorithm: "hmac-sha256",
        keyVersion: keyedDigest.keyVersion,
        digest: prefixedHmacSha256(canonicalJson(toJsonValue(value)), keyedDigest.secret),
      };
    } else if (policy.capture === "content_digest") {
      fields[key] = {
        captureMode: policy.capture,
        digest: prefixedSha256(canonicalJson(toJsonValue(value))),
      };
    } else {
      throw new TypeError(`capture mode ${policy.capture} is not supported by the current governed-change draft helper`);
    }
  }

  return {
    algorithm: "sha256",
    canonicalization: "veritio-json-v1",
    schemaRef: entity.schemaRef,
    fieldSetRef: entity.fieldSetRef,
    fields,
    digest: prefixedSha256(canonicalJson(fields)),
  };
}

/**
 * Creates an edge input and carries the full structured refs in metadata so v1
 * graph records preserve authority even though endpoints use EvidenceEntity.
 */
function draftEdge(
  relation: EvidenceEdgeInput["relation"],
  from: EvidenceRef,
  to: EvidenceRef,
  occurredAt: string,
  scope: EvidenceScope,
): EvidenceEdgeInput {
  return {
    id: `edge_${relation}_${stableId(refKey(from))}_${stableId(refKey(to))}`,
    occurredAt,
    scope,
    from: entityFromRef(from),
    relation,
    to: entityFromRef(to),
    metadata: { fromRef: from, toRef: to },
  };
}

/**
 * Maps EvidenceRef into the current EvidenceEdge endpoint shape while preserving
 * the semantic type as resourceType for UI projections.
 */
function entityFromRef(ref: EvidenceRef): EvidenceEntity {
  assertRef(ref);
  const type = ref.kind === "commit" ? "evidence_commit" : ref.kind;
  const entity: EvidenceEntity = { type: type as EvidenceEntity["type"], id: ref.id, resourceType: ref.type };
  if (ref.kind === "principal" && isGraphActorType(ref.type)) {
    entity.actorType = ref.type;
  }
  return entity;
}

/**
 * Graph actor endpoints intentionally exclude anonymous actors because exported
 * principal refs must remain authority-qualified.
 */
function isGraphActorType(value: string): value is NonNullable<EvidenceEntity["actorType"]> {
  return value === "user" || value === "service" || value === "system" || value === "ai_agent";
}

/**
 * Converts an authority-qualified principal ref into the legacy audit actor
 * shape used by current AuditEvent payloads.
 */
function principalFromRef(ref: EvidenceRef): Principal {
  if (ref.kind !== "principal") {
    throw new TypeError("principal ref is required");
  }
  return { type: ref.type as Principal["type"], id: `${ref.authority}:${ref.id}` };
}

/**
 * Drops undefined metadata fields before audit-event redaction and hashing.
 */
function compactMetadata(value: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [key, item] of Object.entries(value)) {
    if (item !== undefined && item !== null) {
      out[key] = item;
    }
  }
  return out;
}

/**
 * Converts host row values into the JSON value domain accepted by canonicalJson.
 */
function toJsonValue(value: unknown): JsonValue {
  if (value === null || typeof value === "string" || typeof value === "boolean") {
    return value;
  }
  if (typeof value === "number") {
    if (!Number.isFinite(value)) {
      throw new TypeError("field values must be finite JSON numbers");
    }
    return value;
  }
  if (Array.isArray(value)) {
    return value.map(toJsonValue);
  }
  if (value instanceof Date) {
    return value.toISOString();
  }
  if (typeof value === "object" && value !== undefined) {
    const out: JsonObject = {};
    for (const key of Object.keys(value as Record<string, unknown>).sort()) {
      const nested = (value as Record<string, unknown>)[key];
      if (nested !== undefined) {
        out[key] = toJsonValue(nested);
      }
    }
    return out;
  }
  throw new TypeError(`unsupported field value type: ${typeof value}`);
}

/**
 * Normalizes dates to the millisecond ISO representation used by audit events.
 */
function normalizeDate(value: string | Date): string {
  const date = value instanceof Date ? value : new Date(value);
  if (Number.isNaN(date.getTime())) {
    throw new TypeError("occurredAt must be a valid date");
  }
  return date.toISOString();
}

/**
 * Produces a short deterministic token for generated local draft IDs.
 */
function stableId(value: string): string {
  return createHash("sha256").update(value).digest("hex").slice(0, 16);
}

/**
 * Computes the public digest string used in revision commitments.
 */
function prefixedSha256(value: string): string {
  return `sha256:${createHash("sha256").update(value).digest("hex")}`;
}

/**
 * Computes tenant-keyed digests without persisting the raw key or raw value in
 * evidence. Hosts inject key material at the capture boundary.
 */
function prefixedHmacSha256(value: string, secret: string): string {
  return `sha256:${createHmac("sha256", secret).update(value).digest("hex")}`;
}

/**
 * Validates full structured references before they cross SDK boundaries.
 */
function assertRef(ref: EvidenceRef): void {
  assertNonEmpty(ref.authority, "ref.authority");
  assertNonEmpty(ref.kind, "ref.kind");
  assertNonEmpty(ref.type, "ref.type");
  assertNonEmpty(ref.id, "ref.id");
}

/**
 * Requires non-empty strings for public helper inputs.
 */
function assertNonEmpty(value: unknown, field: string): asserts value is string {
  if (typeof value !== "string" || value.trim().length === 0) {
    throw new TypeError(`${field} is required`);
  }
}

File: sdks/python/src/veritio/governed_change.py
from __future__ import annotations

import hashlib
import hmac
from datetime import datetime, timezone
from typing import Any, Callable

from .event import _normalize_json, canonical_json

RESERVED_CONTEXT_KEYS = [
    "authSessionId",
    "authContextId",
    "activityEpisodeId",
    "traceId",
    "correlationId",
    "causationEventId",
    "changeId",
    "capturePolicyId",
    "collectionSource",
]


def ref_key(ref: dict[str, Any]) -> str:
    """Format an authority-qualified evidence reference for stable joins."""
    _assert_ref(ref)
    return f"{ref['authority']}:{ref['kind']}:{ref['type']}:{ref['id']}"


def merge_veritio_metadata(
    caller_metadata: dict[str, Any] | None = None,
    context: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Merge caller metadata with SDK-owned context keys without shadowing."""
    caller = caller_metadata or {}
    for key in RESERVED_CONTEXT_KEYS:
        if key in caller:
            raise TypeError(f"metadata.{key} is reserved by Veritio")

    merged = {key: value for key, value in sorted(caller.items()) if value is not None}
    for key in RESERVED_CONTEXT_KEYS:
        if context and context.get(key) is not None:
            merged[key] = context[key]
    return merged


def define_entity(
    *,
    authority: str,
    entity_type: str,
    schema_ref: str,
    field_set_ref: str,
    identity: Callable[[dict[str, Any]], str],
    fields: dict[str, dict[str, str]],
    lineage_policy: str | None = None,
) -> dict[str, Any]:
    """Register governed entity capture policy at the host boundary.

    Field capture modes IMPLEMENTED by the v1 state-commitment builder (all
    three SDKs): "omit", "content_digest", "keyed_digest", "full"; the
    remaining modes ("randomized_digest", "reference", "redact", "encrypt")
    are RESERVED and fail closed at draft time. ``lineage_policy`` ("linear"
    or "dag") is RESERVED, not yet enforced: no SDK or reference-store branch
    reads it today; host-side enforcement lands with the host-assigned
    revision-ordinal design (docs/review-backlog.md item C).
    """
    _assert_non_empty(authority, "authority")
    _assert_non_empty(entity_type, "type")
    _assert_non_empty(schema_ref, "schemaRef")
    _assert_non_empty(field_set_ref, "fieldSetRef")
    entity = {
        "authority": authority,
        "type": entity_type,
        "schemaRef": schema_ref,
        "fieldSetRef": field_set_ref,
        "identity": identity,
        "fields": fields,
    }
    if lineage_policy:
        entity["lineagePolicy"] = lineage_policy
    return entity


def governed_revision_id(entity_type: str, entity_id: str, state_digest: str, change_id: str) -> str:
    """Derive the deterministic revision id for a governed-state commitment.

    Content-addressed by the state digest AND scoped by the producing change
    (first 8 hex chars of sha256(change_id)), so a rollback that restores a
    byte-identical earlier state still yields a DISTINCT revision id while
    replaying the same change stays idempotent. Byte-identical across the
    TS/Python/Go SDKs (spec/conformance/governed-revision-id.json). The design
    target remains a host-assigned ordinal suffix; hosts that assign ordinals
    supersede this derivation.
    """
    digest12 = state_digest[len("sha256:") : len("sha256:") + 12]
    change8 = hashlib.sha256(change_id.encode("utf-8")).hexdigest()[:8]
    return f"rev_{entity_type}_{entity_id}_{digest12}_{change8}"


def create_governed_change_draft(input_change: dict[str, Any]) -> dict[str, Any]:
    """Create current-protocol audit events and edges for a governed change."""
    scope = input_change["scope"]
    _assert_non_empty(scope.get("tenantId"), "scope.tenantId")
    entity = input_change["entity"]
    after = input_change["after"]
    entity_ref = _entity_ref(entity, after)
    change = input_change["change"]
    activity = input_change["activity"]
    producer = input_change["producer"]
    _assert_ref(change["initiatedBy"])
    _assert_ref(activity["performedBy"])
    _assert_ref(producer)
    try:
        occurred_at = _normalize_datetime(input_change["occurredAt"])
    except (ValueError, TypeError):
        # Raise the same typed message as the TS/Go SDKs instead of leaking the
        # stdlib parser's text (which echoes the raw input); occurredAt is hashed
        # evidence, so the cross-language error contract must match.
        raise ValueError("occurredAt must be a valid date") from None

    change_ref = {"authority": "veritio", "kind": "change", "type": change["type"], "id": change["id"]}
    activity_ref = {"authority": "veritio", "kind": "activity", "type": activity["type"], "id": activity["id"]}
    # Use ONLY a caller-supplied parent revision; never fabricate a placeholder
    # rev_..._previous (mirrors the TS/Go SDKs). A synthetic parent asserts a
    # false derived_from edge to a revision that never existed and feeds the host
    # store an optimistic-concurrency token its real head can never match.
    expected_parent = input_change.get("expectedParentRevisionRef")
    if expected_parent:
        _assert_ref(expected_parent)
    parent_ref = expected_parent if input_change.get("before") else None
    state_commitment = _state_commitment(entity, after, input_change.get("digestKeys") or {})
    revision_ref = {
        "authority": "veritio",
        "kind": "revision",
        "type": entity["type"],
        "id": governed_revision_id(entity["type"], entity_ref["id"], state_commitment["digest"], change["id"]),
    }
    revision = {
        "ref": revision_ref,
        "entity": entity_ref,
        "parents": [parent_ref] if parent_ref else [],
        "stateCommitment": state_commitment,
        "changedPaths": sorted(input_change.get("changedPaths") or []),
        "generatedBy": activity_ref,
    }
    if input_change.get("capturePolicyRef"):
        revision["capturePolicyRef"] = input_change["capturePolicyRef"]

    metadata = merge_veritio_metadata(input_change.get("metadata"), input_change.get("context"))
    capture_assurance = {
        "captureMethod": "transactional_outbox",
        "mutationBinding": input_change.get("mutationBinding") or "not_transaction_bound",
    }
    common = {
        "scope": scope,
        "occurredAt": occurred_at,
        "purpose": "change_provenance",
        "dataCategories": ["source_reference"],
        "retention": "change_1y",
    }
    events = [
        {
            **common,
            "id": f"evt_change_declared_{change['id']}",
            "actor": _principal_from_ref(change["initiatedBy"]),
            "action": "change.declared",
            "target": {"type": "change", "id": change["id"]},
            "metadata": _compact({
                **metadata,
                "recordType": "change.declared",
                "recordAuthority": change_ref["authority"],
                "producer": producer,
                "initiatedBy": change["initiatedBy"],
                "changeType": change["type"],
                "idempotencyKeyHash": input_change["idempotencyKeyHash"],
                "capturePolicyRef": input_change.get("capturePolicyRef"),
                "authorizationAssertionRef": change.get("authorizationAssertionRef"),
                "delegationAssertionRef": change.get("delegationAssertionRef"),
                "captureAssurance": capture_assurance,
            }),
        },
        {
            **common,
            "id": f"evt_activity_recorded_{activity['id']}",
            "actor": _principal_from_ref(activity["performedBy"]),
            "action": "activity.recorded",
            "target": {"type": "activity", "id": activity["id"]},
            "metadata": _compact({
                **metadata,
                "recordType": "activity.recorded",
                "recordAuthority": activity_ref["authority"],
                "producer": producer,
                "performedBy": activity["performedBy"],
                "activityType": activity["type"],
                "idempotencyKeyHash": input_change["idempotencyKeyHash"],
                "captureAssurance": capture_assurance,
            }),
        },
        {
            **common,
            "id": f"evt_entity_revision_{revision_ref['id']}",
            "actor": _principal_from_ref(producer),
            "action": "entity.revision.created",
            "target": {"type": entity["type"], "id": entity_ref["id"]},
            "metadata": _compact({
                **metadata,
                "recordType": "entity.revision",
                "recordAuthority": revision_ref["authority"],
                "producer": producer,
                "idempotencyKeyHash": input_change["idempotencyKeyHash"],
                "veritio": {"revision": revision},
                "captureAssurance": capture_assurance,
            }),
        },
    ]
    edges = [
        _draft_edge("has_activity", change_ref, activity_ref, occurred_at, scope),
        _draft_edge("has_output", change_ref, revision_ref, occurred_at, scope),
        _draft_edge("performed_by", activity_ref, activity["performedBy"], occurred_at, scope),
        _draft_edge("generated", activity_ref, revision_ref, occurred_at, scope),
    ]
    if parent_ref:
        edges.append(_draft_edge("derived_from", revision_ref, parent_ref, occurred_at, scope))

    outbox_entry = {
        "schemaVersion": "2026-06-23",
        "mutationBinding": capture_assurance["mutationBinding"],
        "records": events,
        "edges": edges,
    }
    if parent_ref:
        outbox_entry["expectedParentRevisionRef"] = parent_ref

    return {
        "changeRef": change_ref,
        "activityRef": activity_ref,
        "entityRef": entity_ref,
        "revision": revision,
        "events": events,
        "edges": edges,
        "outboxEntry": outbox_entry,
    }


def _state_commitment(entity: dict[str, Any], row: dict[str, Any], digest_keys: dict[str, Any]) -> dict[str, Any]:
    """Apply field capture policy before evidence leaves the host process."""
    fields: dict[str, Any] = {}
    for key in sorted(entity["fields"].keys()):
        policy = entity["fields"][key]
        mode = policy.get("capture")
        if mode == "omit" or key not in row:
            continue
        value = row[key]
        if mode == "full":
            fields[key] = _normalize_json(value)
        elif mode == "keyed_digest":
            keyed_digest = digest_keys.get("keyedDigest")
            if not keyed_digest:
                raise TypeError("digestKeys.keyedDigest is required for keyed_digest fields")
            _assert_non_empty(keyed_digest.get("keyVersion"), "digestKeys.keyedDigest.keyVersion")
            _assert_non_empty(keyed_digest.get("secret"), "digestKeys.keyedDigest.secret")
            fields[key] = {
                "algorithm": "hmac-sha256",
                "keyVersion": keyed_digest["keyVersion"],
                "digest": _prefixed_hmac_sha256(canonical_json(value), keyed_digest["secret"]),
            }
        elif mode == "content_digest":
            fields[key] = {"captureMode": mode, "digest": _prefixed_sha256(canonical_json(value))}
        else:
            raise TypeError(f"capture mode {mode} is not supported by the current governed-change draft helper")

    return {
        "algorithm": "sha256",
        "canonicalization": "veritio-json-v1",
        "schemaRef": entity["schemaRef"],
        "fieldSetRef": entity["fieldSetRef"],
        "fields": fields,
        "digest": _prefixed_sha256(canonical_json(fields)),
    }


def _draft_edge(relation: str, from_ref: dict[str, Any], to_ref: dict[str, Any], occurred_at: str, scope: dict[str, Any]) -> dict[str, Any]:
    """Build an EvidenceEdge input while preserving structured refs in metadata."""
    return {
        "id": f"edge_{relation}_{_stable_id(ref_key(from_ref))}_{_stable_id(ref_key(to_ref))}",
        "occurredAt": occurred_at,
        "scope": scope,
        "from": _entity_from_ref(from_ref),
        "relation": relation,
        "to": _entity_from_ref(to_ref),
        "metadata": {"fromRef": from_ref, "toRef": to_ref},
    }


def _entity_ref(entity: dict[str, Any], row: dict[str, Any]) -> dict[str, str]:
    """Resolve a host row into an authority-qualified entity ref."""
    entity_id = entity["identity"](row)
    _assert_non_empty(entity_id, "entity.id")
    return {"authority": entity["authority"], "kind": "entity", "type": entity["type"], "id": entity_id}


def _entity_from_ref(ref: dict[str, Any]) -> dict[str, str]:
    """Map EvidenceRef into the current EvidenceEdge endpoint shape."""
    endpoint_type = "evidence_commit" if ref["kind"] == "commit" else ref["kind"]
    entity = {"type": endpoint_type, "id": ref["id"], "resourceType": ref["type"]}
    if ref["kind"] == "principal" and ref["type"] in {"user", "service", "system", "ai_agent"}:
        entity["actorType"] = ref["type"]
    return entity


def _principal_from_ref(ref: dict[str, Any]) -> dict[str, str]:
    """Convert a principal ref into the legacy AuditEvent actor shape."""
    if ref.get("kind") != "principal":
        raise TypeError("principal ref is required")
    return {"type": ref["type"], "id": f"{ref['authority']}:{ref['id']}"}


def _compact(value: dict[str, Any]) -> dict[str, Any]:
    """Drop absent optional fields before event normalization redacts metadata."""
    return {key: nested for key, nested in value.items() if nested is not None}


def _normalize_datetime(value: str | datetime) -> str:
    """Normalize date strings and datetimes to UTC millisecond ISO strings."""
    if isinstance(value, datetime):
        date = value
    else:
        date = datetime.fromisoformat(value.replace("Z", "+00:00"))
    if date.tzinfo is None:
        date = date.replace(tzinfo=timezone.utc)
    date = date.astimezone(timezone.utc)
    return date.isoformat(timespec="milliseconds").replace("+00:00", "Z")


def _stable_id(value: str) -> str:
    """Produce a short deterministic ID token for generated edge IDs."""
    return hashlib.sha256(value.encode("utf-8")).hexdigest()[:16]


def _prefixed_sha256(value: str) -> str:
    """Compute the prefixed SHA-256 digest string used by commitments."""
    return "sha256:" + hashlib.sha256(value.encode("utf-8")).hexdigest()


def _prefixed_hmac_sha256(value: str, secret: str) -> str:
    """Compute a host-keyed digest without persisting raw key or raw value."""
    return "sha256:" + hmac.new(secret.encode("utf-8"), value.encode("utf-8"), hashlib.sha256).hexdigest()


def _assert_ref(ref: dict[str, Any]) -> None:
    """Validate the authority-qualified reference shape."""
    _assert_non_empty(ref.get("authority"), "ref.authority")
    _assert_non_empty(ref.get("kind"), "ref.kind")
    _assert_non_empty(ref.get("type"), "ref.type")
    _assert_non_empty(ref.get("id"), "ref.id")


def _assert_non_empty(value: Any, field: str) -> None:
    """Require non-empty public string fields."""
    if not isinstance(value, str) or not value.strip():
        raise TypeError(f"{field} is required")
